flows/engine/assets.go

```diff
@@ -3,90 +3,142 @@ package engine
 	}
-	return nil, fmt.Errorf("unable to find flow with UUID: %s", uuid)
+	return flow, nil
 }
 
-func (a *assets) GetChannel(uuid flows.ChannelUUID) (flows.Channel, error) {
-	channel, exists := a.channels[uuid]
-	if exists {
-		return channel, nil
+func (m *assetManager) GetChannel(uuid flows.ChannelUUID) (flows.Channel, error) {
+	asset, err := m.requestAsset(flows.AssetUUID(uuid), "channel")
+	if err != nil {
+		return nil, err
+	}
+	channel, isType := asset.(flows.Channel)
+	if !isType {
+		return nil, fmt.Errorf("unable to find channel with UUID '%s'", uuid)
 	}
-	return nil, fmt.Errorf("unable to find channel with UUID: %s", uuid)
+	return channel, nil
 }
 
 //------------------------------------------------------------------------------------------
 // JSON Encoding / Decoding
 //------------------------------------------------------------------------------------------
 
-type assetsEnvelope struct {
-	Flows    []json.RawMessage `json:"flows"               validate:"required"`
-	Channels []json.RawMessage `json:"channels,omitempty"`
+type assetEnvelope struct {
+	Type    flows.AssetType  `json:"type"    validate:"required"`
+	UUID    *flows.AssetUUID `json:"uuid"    validate:"omitempty,uuid"`
+	Content *json.RawMessage `json:"content" validate:"omitempty"`
+	URL     string           `json:"url"     validate:"omitempty,url"`
 }
 
-func ReadAssets(data json.RawMessage) (flows.Assets, error) {
-	var envelope assetsEnvelope
-	err := utils.UnmarshalAndValidate(data, &envelope, "assets")
-	if err != nil {
-		return nil, err
+func (m *assetManager) IncludeAssets(data json.RawMessage) error {
+	var raw []json.RawMessage
+	if err := json.Unmarshal(data, &raw); err != nil {
+		return err
 	}
 
-	channelList := make([]flows.Channel, len(envelope.Channels))
-	flowList := make([]flows.Flow, len(envelope.Flows))
+	envelopes := make([]assetEnvelope, len(raw))
+	for e := range raw {
+		if err := utils.UnmarshalAndValidate(raw[e], &envelopes[e], "asset"); err != nil {
+			return err
+		}
+	}
 
-	for c := range envelope.Channels {
-		channel, err := flows.ReadChannel(envelope.Channels[c])
-		if err != nil {
-			return nil, err
+	nonLazy := make([]flows.Asset, 0)
+
+	for _, envelope := range envelopes {
+		var asset flows.Asset
+
+		if envelope.Content != nil {
+			var err error
+			switch envelope.Type {
+			case flows.AssetTypeFlow:
+				asset, err = definition.ReadFlow(*envelope.Content)
+			case flows.AssetTypeChannel:
+				asset, err = flows.ReadChannel(*envelope.Content)
+			default:
+				err = fmt.Errorf("Invalid asset type: %s", envelope.Type)
+			}
+			if err != nil {
+				return err
+			}
+
+			nonLazy = append(nonLazy, asset)
+			m.AddAsset(asset, envelope.URL)
+		} else {
+			m.AddLazyAsset(envelope.Type, *envelope.UUID, envelope.URL)
 		}
-		channelList[c] = channel
 	}
 
-	for f := range envelope.Flows {
-		flow, err := definition.ReadFlow(envelope.Flows[f])
-		if err != nil {
-			return nil, err
+	// any non-lazy assets can be validated now
```

Comment:
This is a bit weird - say someone passes in a flow and a subflow which the first flow requires. If we validate each asset before adding it, the first one will trigger another request for the subflow... so I'm adding them all to the asset cache and then validating them.